django/db/models/query.py

```diff
@@ -1581,8 +1582,19 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
 
         if single:
             val = vals[0] if vals else None
-            to_attr = to_attr if as_attr else cache_name
-            setattr(obj, to_attr, val)
+
+            if as_attr:
+                # A to_attr has been given for the prefetch
+                setattr(obj, to_attr, val)
+            elif is_descriptor:
+                # cache_name points to a field name in obj
+                # this field is a descriptor for a related object.
+                setattr(obj, cache_name, val)
+            else:
+                # No to_attr has been given for this prefetch operation
+                # and the cache_name does not point to a descriptor.
+                # store the value of the field directly in the object's field cache.
+                obj._state.cache[cache_name] = val
```

Comment:
Thanks for the feedback.
I'll have a deeper look to see how this can be improved.